compat/openssl.h

```diff
@@ -0,0 +1,241 @@
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+/* OpenSSL API changed dramatically between 1.0.2 and 1.1.0, and
+ * compatibility was broken.  Most of the structures became opaque,
+ * and access functions were created.  There's no (safe) way to
+ * access the struct members any more, so the solution is to use
+ * the new API in the main code, and add the functions for older
+ * versions in compat/openssl.h.
+ * Once all the supported library versions use the new API, the shim
+ * can be dropped.
+ */
+
+#ifndef OPENSSL_COMPAT_H
+#define OPENSSL_COMPAT_H
+
+#if !USE_OPENSSL
+#error compat/openssl.h depends on USE_OPENSSL
+#endif
+
+#if defined(__cplusplus)
+
+#if HAVE_OPENSSL_ASN1_H
+#include <openssl/asn1.h>
+#endif
+#if HAVE_OPENSSL_BIO_H
+#include <openssl/bio.h>
+#endif
+#if HAVE_OPENSSL_DH_H
+#include <openssl/dh.h>
+#endif
+#if HAVE_OPENSSL_EVP_H
+#include <openssl/evp.h>
+#endif
+#if HAVE_OPENSSL_LHASH_H
+#include <openssl/lhash.h>
+#endif
+#if HAVE_OPENSSL_SSL_H
+#include <openssl/ssl.h>
+#endif
+#if HAVE_OPENSSL_X509_H
+#include <openssl/x509.h>
+#endif
+
+extern "C" {
+
+#if !HAVE_LIBCRYPTO_ASN1_STRING_GET0_DATA
+inline const unsigned char *
+ASN1_STRING_get0_data(const ASN1_STRING *x)
+{
+    return x->data;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_BIO_GET_DATA
+inline void *
+BIO_get_data(BIO *table)
+{
+    return table->ptr;
+}
+
+inline void
+BIO_set_data(BIO *table, void *data)
+{
+   table->ptr = data;
+}
+
+inline void
+BIO_set_init(BIO *table, int init)
+{
+   table->init = init;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_BIO_GET_INIT
+inline int
+BIO_get_init(BIO *table)
+{
+    return table->init;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_DH_UP_REF // OpenSSL 1.1 API
+#if defined(CRYPTO_LOCK_DH) // OpenSSL 1.0 API
+inline int
+DH_up_ref(DH *t)
+{
+    if (t && (CRYPTO_add(&t->references, 1, CRYPTO_LOCK_DH) > 1))
+        return 1;
+    return 0;
+}
+#else
+#error missing both OpenSSL API features DH_up_ref (v1.1) and CRYPTO_LOCK_DH (v1.0)
+#endif /* OpenSSL 1.0 CRYPTO_LOCK_DH */
+#endif /* OpenSSL 1.1 DH_up_ref */
+
+#if !HAVE_LIBCRYPTO_EVP_PKEY_GET0_RSA
+inline RSA *
+EVP_PKEY_get0_RSA(EVP_PKEY *pkey)
+{
+    if (pkey->type != EVP_PKEY_RSA)
+        return nullptr;
+    return pkey->pkey.rsa;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_EVP_PKEY_UP_REF
+#if defined(CRYPTO_LOCK_EVP_PKEY) // OpenSSL 1.0
+inline int
+EVP_PKEY_up_ref(EVP_PKEY *t)
+{
+    if (t && (CRYPTO_add(&t->references, 1, CRYPTO_LOCK_EVP_PKEY)) > 1)
+        return 1;
+    return 0;
+}
+
+#else
+#error missing both OpenSSL API features EVP_PKEY_up_ref (v1.1) and CRYPTO_LOCK_EVP_PKEY (v1.0)
+#endif /* OpenSSL 1.0 CRYPTO_LOCK_EVP_PKEY */
+#endif /* OpenSSL 1.1 EVP_PKEY_up_ref */
+
+#if !HAVE_LIBCRYPTO_OPENSSL_LH_STRHASH
+#define OPENSSL_LH_delete lh_delete
+#define OPENSSL_LH_strhash lh_strhash
+#endif
+
+inline void SQUID_OPENSSL_init_ssl(void)
```

Comment:
I've ran the script, and it changed the indentation of the shim functions to conform to the `extern "C"` braces.  Since I found more code that followed this style, I've committed the change.

Reply:
I think you did the right thing here: We should accept astyle imperfect formatting -- tuning it is outside this PR scope. It is perhaps debatable whether we should have such a huge `extern "C"` block, but that is a separate question (that I am not asking on this change thread).